ADR/0015-integration-service-two-phase-architecture.md

```diff
@@ -0,0 +1,109 @@
+# 15. The Two-phase Architecture of the Integration Service
+
+Date Documented: 2023-02-09
+
+## Status
+
+In review
+
+## Context
+
+The Integration Service is in charge of running integration test pipelines by executing
+the Tekton pipeline for each user-defined IntegrationTestScenario.
+The main payloads that are being tested are Snapshots which contain references to all
+Components that belong to the Application, along with their images.
+
+One problem faced by the integration service is caused by the fact that testing an
+application with Tekton pipelines takes time. During the course of testing an application,
+multiple component builds can happen in quick succession, leading to a potential race
+condition between different Snapshots that are created for each of those builds.
+This would primarily manifest itself by two Snapshots getting promoted in quick
+succession but neither of them having the latest images contained within them.
+
+## Decision
+
+In order to protect against the race conditions, the integration service will leverage a
+two phase approach to testing. The phases will consist of a Component phase that will
+always be executed and the optional Composite phase which would come into play only when
+race conditions between component builds are detected.
+
+### Component phase
+
+When a single component image is built, the Integration Service tests the application
+by creating a Snapshot. All Components with their images from the Global Candidate List
+are included within the Snapshot and then the Component that was newly built is
+updated/overwritten to complete the Snapshot creation.
+
+After all test pipelines for the Snapshot finish successfully, the Integration service
+updates the Global Candidate List with the newly built Component image and checks if it
+can promote the Component Snapshot. If the Global Candidate List for other Components
+doesn't match the rest of the Component Snapshot contents, its status is marked as
+invalid and the testing goes into the Composite phase.
+Otherwise, the Component Snapshot is promoted according to user preferences.
+
+![](../diagrams/ADR-0015/component-phase.jpg)
+
+### Composite phase
+
+The Composite phase is used when the Global Candidate List changes while testing a
+Snapshot in the Component phase. The Integration service checks this before promoting
+a Snapshot at the end of the Component phase.
```

Comment:
I don't think this last sentence is needed. Otherwise, I think it reads well!

Reply:
Yep, I can see it now too, removed it.